components/web_discovery/browser/document_extractor/rs/src/lib.rs

```diff
@@ -0,0 +1,124 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::collections::HashMap;
+
+use cxx::{CxxString, CxxVector};
+use html5ever::tree_builder::TreeSink;
+use kuchikiki::{
+    iter::{Descendants, Elements, Select},
+    parse_html,
+    traits::TendrilSink,
+};
+
+#[cxx::bridge(namespace = "rust_document_extractor")]
+mod ffi {
+    pub struct SelectAttributeRequest {
+        /// An optional selector for an element within the current selected element.
+        /// The attribute will be retrieved from the embedded element.
+        /// If not needed, an empty string should be provided.
+        pub sub_selector: String,
+        /// Arbitrary ID used for storing the scraped result.
+        pub key: String,
+        /// Name of the attribute to scrape.
+        pub attribute: String,
+    }
+
+    pub struct SelectRequest {
+        /// The DOM selector for the element to scrape.
+        pub root_selector: String,
+        /// Scrape requests for the selected element.
+        pub attribute_requests: Vec<SelectAttributeRequest>,
+    }
+
+    pub struct AttributePair {
+        /// Arbitrary ID for the scraped result.
+        pub key: String,
+        /// The scraped value. Will be empty if attribute is not available.
+        pub value: String,
+    }
+
+    pub struct AttributeResult {
+        /// The DOM selector for the scraped element.
+        pub root_selector: String,
+        /// A list of arbitrary IDs and scraped value pairs.
+        pub attribute_pairs: Vec<AttributePair>,
+    }
+
+    extern "Rust" {
+        /// Extracts DOM attributes from the result of a double fetch.
+        fn query_element_attributes(
+            html: &CxxString,
+            requests: &CxxVector<SelectRequest>,
+        ) -> Vec<AttributeResult>;
+    }
+}
+
+use ffi::*;
+
+const TEXT_CONTENT_ATTRIBUTE_NAME: &str = "textContent";
+
+fn extract_attributes_from_nodes(
```

Comment:
can't we use the axtree which would also take the place of the document extrator?

Reply:
The crate is used to process the page contents from the anonymous request made in the browser process. If there is a way to construct an AxTree from a string, would it be easy to do this in a sandboxed process? if not, wouldn't that violate rule of 2?